[JSC] Apply SIMD string scanning for sloppy JSON too

Constellation opened the pull request:
#### 884ee2aecdc035f729120cdd61137381581530ac
<pre>
[JSC] Apply SIMD string scanning for sloppy JSON too
<a href="https://bugs.webkit.org/show_bug.cgi?id=296847">https://bugs.webkit.org/show_bug.cgi?id=296847</a>
<a href="https://rdar.apple.com/157377100">rdar://157377100</a>

Reviewed by Yijia Huang.

After [1] proposal gets integrated into ECMA-262, JSON v.s. non-JSON
string scanning can be much simpler. Now this patch revisits the
predicate for Sloppy JSON string scanning and integrating SIMD fast
scanning for strings.

[1]: <a href="https://github.com/tc39/proposal-json-superset">https://github.com/tc39/proposal-json-superset</a>

* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::isSafeStringCharacter):
(JSC::reviverMode&gt;::Lexer::lexString):

Canonical link: <a href="https://commits.webkit.org/298208@main">https://commits.webkit.org/298208@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c2673dfde085b7884a1f9c62606149b2d9b3116

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114637 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34383 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24846 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120803 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/65359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/48eb7863-b43f-46bc-a8cb-4abfc4c8fc80) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35011 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42943 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/87138 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/65359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/86033b24-f8ee-455e-a01b-727f08c78ffe) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117585 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27895 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102955 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67526 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27082 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21079 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/64480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/107029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/97275 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21189 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124005 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/113200 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41649 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/31101 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/95953 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42026 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99140 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95734 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40901 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18746 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18365 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41527 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47039 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/137414 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41113 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36744 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44420 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/42863 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8a10fd0cf66b63ce500032b0283649fcc4dadb57. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48867

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bd2a533ec3d3c5584ae5aae4cef7bf9635cfee5f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48867

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Apply SIMD string scanning for sloppy JSON too
https://bugs.webkit.org/show_bug.cgi?id=296847
rdar://157377100

Reviewed by Yijia Huang.

After [1] proposal gets integrated into ECMA-262, JSON v.s. non-JSON
string scanning can be much simpler. Now this patch revisits the
predicate for Sloppy JSON string scanning and integrating SIMD fast
scanning for strings.

[1]: https://github.com/tc39/proposal-json-superset

* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::isSafeStringCharacter):
(JSC::reviverMode>::Lexer::lexString):

Canonical link: https://commits.webkit.org/298208@main
